### PR TITLE
changes link file on Android to MainApplication.java for 0.29 update

### PR DESCRIPTION
**Pull request by GantMan**
rnpm aka `react-native link` is broken with Android 0.29 - #8603 

This gets it back to working again by checking for new MyApplication.java file, and curtailing the path when needed.

### CODE REVIEW

**Comment by facebook-github-bot-5** [comment]
By analyzing the blame information on this pull request, we identified @grabbou to be a potential reviewer.

**Comment by GantMan** [comment]
note, the methodology for checking the file is the latest method: http://stackoverflow.com/questions/4482686/check-synchronously-if-file-directory-exists-in-node-js

**Comment by satya164** [comment]
cc @Kureev

**Comment by Kureev** [comment]
Hi @GantMan! Thanks for proposing this PR! I have a few small questions in order to process it faster :)
As I see, you're checking for a MainApplication.java and "roll back" to MainActivity.java in a case if it's not found. But taking into account that rnpm is now a part of react-native cli, I think we don't have any cases to support any previous versions. Does it make any sense?

**Comment by GantMan** [comment]
Hi @Kureev!   My guess use case where someone would care is this:  "I created project X in RN 0.28 and I created project Y in 0.29".   I should be able to still work on my 0.28 project and even use `react-native link` to add new features to it.   It shouldn't break functionality on my non-updated project.   I might have good reason for not bringing it to 0.29 just yet.

**Comment by satya164** [comment]
May be support old versions for few releases and then drop the support.

**Comment by GantMan** [comment]
Yeah, ideal would be 3 phases.

1. this goes through as hotfix.
2. deprecate, and warn
3. remove

I'm happy to help with all 3 steps.

**Comment by Kureev** [comment]
@GantMan For your case with 0.28 & 0.29: if I understand correctly, app scaffolded by 0.28 will have a react-native@0.28 with current `react-native link` implementation under the hood. Same is fair for 0.29: it'll have a new 0.29 react-native with new `link` for MainApplication. The only controversial point is about `react-native upgrade`. Does it convert your MainActivity to MainApplication? If so, then I don't see a reason why we need prev. versions. 

If 0.28 to 0.29 migration by using `react-native upgrade` wouldn't rename MainActivity to MainApplication, then we need to add it as a part of the migration (which is a desired behavior imo) or support N releases behind the last one (for `react-native link`).

File: local-cli/rnpm/core/src/config/android/index.js [hunk 1]
```diff
@@ -33,11 +33,28 @@ exports.projectConfig = function projectConfigAndroid(folder, userConfig) {
...
     userConfig.mainActivityPath || `src/main/java/${packageFolder}/MainActivity.java`
   );
 
+  // If MainApplication.java exists, use that instead
+  const mainApplicationPath = path.join(
+    sourceDir,
+    userConfig.mainApplicationPath || `src/main/java/${packageFolder}/MainApplication.java`
+  );
+
+  fs.accessSync(mainApplicationPath, fs.F_OK, function(err) {
```
Comment (Kureev) [inline local-cli/rnpm/core/src/config/android/index.js]: There should be no callback (it's Sync):
```js
try {
  fs.accessSync(mainApplicationPath, fs.F_OK);
} catch (e) {
  //...error
}
//...success
```

File: local-cli/rnpm/core/src/config/android/index.js [hunk 2]
```diff
@@ -33,11 +33,28 @@ exports.projectConfig = function projectConfigAndroid(folder, userConfig) {
...
+
+  mainActivityPath = path.join(
     sourceDir,
     userConfig.mainActivityPath || `src/main/java/${packageFolder}/MainActivity.java`
   );
 
+  // If MainApplication.java exists, use that instead
+  const mainApplicationPath = path.join(
+    sourceDir,
+    userConfig.mainApplicationPath || `src/main/java/${packageFolder}/MainApplication.java`
```
Comment (Kureev) [inline local-cli/rnpm/core/src/config/android/index.js]: I'd rename it from MainActivity/MainApplication to something more generic like `mainFilePath` or whatever is better. Don't like the idea of having `mainApplicationPath` along with `mainActivityPath`.

File: local-cli/rnpm/core/src/config/android/index.js [hunk 3]
```diff
@@ -33,11 +33,28 @@ exports.projectConfig = function projectConfigAndroid(folder, userConfig) {
   const packageFolder = userConfig.packageFolder ||
     packageName.replace(/\./g, path.sep);
 
-  const mainActivityPath = path.join(
+  // variable misnomer - Actually set to application path for 0.29+
+  let mainActivityPath = null;
```
Comment (Kureev) [inline local-cli/rnpm/core/src/config/android/index.js]: nit: maybe direct assign would fit better?

```js
let mainFilePath = path.join(...);
```

**Comment by satya164** [comment]
> if I understand correctly, app scaffolded by 0.28 will have a react-native@0.28 with current react-native link implementation under the hood

@Kureev is right. I forgot that the cli just forwards the commands to locally installed `react-native`. So we should not need check for older versions in the code.

**Comment by GantMan** [comment]
I will update accordingly, you'll see code soon :)

**Comment by GantMan** [comment]
@Kureev  and @satya164 -  Thanks for the feedback.  

I wasn't sure what the userConfig variable should be, but I figured consistency with the internal name would be best.   We don't want users setting `userConfig.mainActivityPath` and thinking that's correct.

I did a quick test, and `upgrade` from 0.28-> 0.29 does NOT create the application file.  It's set as manual steps in the release: https://github.com/facebook/react-native/releases/tag/v0.29.0

**Comment by satya164** [comment]
@GantMan Can you also update the commit title to reflect the actual change?

**Comment by satya164** [comment]
@GantMan Yes, `react-native upgrade` doesn't work for 0.28 - 0.29 and it'll probably be very complex to fix this since we're using yeoman for this.

**Comment by GantMan** [comment]
@satya164 - title is updated.

Hah yes, yeoman is quite tricky.  Fortunately it's a separate concern. I was just checking for @Kureev as he seemed to inquire specifically on this upgrade path.   Let me know if there's more on this ticket that needs attention.

**Comment by Kureev** [comment]
I don't think there is anything else. Seems you covered everything, @GantMan 😉

**Comment by Kureev** [comment]
So, I'm going to check it on my local machine today and if everything is fine, I'm about to merge it

**Comment by grabbou** [comment]
Great work! I like it. Another thing is that we remove `mainActivityPath` now, which could potentially break plugins, but since there's no need to use main activity anymore on the Android side, I think it's fine.

**Comment by Kureev** [comment]
@facebook-github-bot shipit

**Comment by Kureev** [comment]
@grabbou we'll notify users in the release notes 👍 Good catch!

**Comment by facebook-github-bot-7** [comment]
Thanks for importing. If you are an FB employee go to [Phabricator](https://our.intern.facebook.com/intern/opensource/github/pull_request/1666672260322360/int_phab) to review.

**Comment by grabbou** [comment]
@mkonicek shall we release a 0.29.1 and also cherry-pick that into 0.30rc? @Kureev

**Comment by satya164** [comment]
@grabbou There was a bug in `InteractionManager` which we should also include. cc @sahrens

**Comment by grabbou** [comment]
@satya164 into the 0.29 or 0.30 branch?

**Comment by satya164** [comment]
@grabbou Both

**Comment by grabbou** [comment]
Ping me on Messagner later just in case

**Comment by sahrens** [comment]
Sorry, still waiting on review. If you need to patch locally, the fix is a one-liner, just change the filter call in `TaskQueue#cancelTasks` from `.filter((queue) => queue.tasks.length > 0);` to:

    .filter((queue, idx) => (queue.tasks.length > 0 || idx === 0));

**Comment by Kureev** [comment]
We missed one change here! Since we renamed `mainActivityPath` to `mainFilePath`, it should be also changed in the `userConfig` object. Want to hotfix, @GantMan?

**Comment by Kureev** [comment]
Seems this is a real scope of the change: https://github.com/facebook/react-native/search?utf8=%E2%9C%93&q=mainActivityPath. Hopefully, it's just about `find . -type f -exec sed -i 's/mainActivityPath/mainFilePath/g' {} +`

**Comment by Kureev** [comment]
see referenced PR for more info.

**Comment by rclai** [comment]
Glad to know I wasn't insane. When are you guys going to finish fixing this?

**Comment by GantMan** [comment]
Looks good @Kureev.   

@rclai - look at PR #8807 

Let me know if I can help any further.

**Pull request closed without merging**